### PR TITLE
STYLE enable ruff PLW2901

### LINKED ISSUE #51708: STYLE enable ruff PLW2901
Task is:

1. check the [contributing guide](https://pandas.pydata.org/docs/dev/development/contributing_environment.html) for how to set up your environment
2. remove `PLW2901` from

https://github.com/pandas-dev/pandas/blob/2c10d93b8b2ccee79434c251ac9c1cd94513cc1e/pyproject.toml#L272-L273

3. run `pre-commit run ruff --all-files`
4. fixup the errors it flags - see https://beta.ruff.rs/docs/rules/redefined-loop-name/ for more info on this error
5. stage, commit, push, open pull request

### PR DESCRIPTION
**Pull request by ErdiTk**
- [ ] closes #51708
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added relative entries in "to be gradually enabled" inside pyproject.toml.

### CODE REVIEW

File: doc/source/conf.py [hunk 1]
```diff
@@ -101,20 +101,20 @@
         reldir = os.path.relpath(dirname, source_path)
         for fname in fnames:
             if os.path.splitext(fname)[-1] in (".rst", ".ipynb"):
-                fname = os.path.relpath(os.path.join(dirname, fname), source_path)
+                fname_new = os.path.relpath(os.path.join(dirname, fname), source_path)
```
Comment (MarcoGorelli) [inline doc/source/conf.py:104]: this is on the right track, but in general, unless there's a meaningful name to give, slight preference for prefixing with an underscore (`_fname`)

Comment (MarcoGorelli) [inline doc/source/conf.py:104]: thanks for working on this - naming variables is hard, I'll take a closer look this week

Comment (ErdiTk) [reply doc/source/conf.py:104]: No problem, I will apply the rule of prefixing with an "_". My bad I should've asked earlier if there was a better standard way.
Yesterday I had some spare time and completed the gradual enabling of "pandas/plotting/*" and "pandas/tests/*". Currently halfway for completing "pandas/io/*". Is it fine if I put everything in a single pull request? I think I can manage to make all the changes.

Comment (MarcoGorelli) [reply doc/source/conf.py:104]: separate would be easier to review, would prefer that if possible :)

Comment (MarcoGorelli) [reply doc/source/conf.py:104]: >  I will apply the rule of prefixing with an "_".

actually, please wait, sorry - we should probably favour descriptive names

Comment (MarcoGorelli) [reply doc/source/conf.py:104]: maybe this one could be `rel_fname`?

Comment (ErdiTk) [reply doc/source/conf.py:104]: I understand the inner battle between descriptive names and verbosity :D.
I will will rethink a bit more about the name and update this PR ok? 
Then for the other sections of "tp be gradually enabled" I can generate other PRs from different branches for it.

File: pandas/util/version/__init__.py [hunk 2]
```diff
@@ -253,16 +253,16 @@ def is_devrelease(self) -> bool:
 
 def _parse_version_parts(s: str) -> Iterator[str]:
     for part in _legacy_version_component_re.split(s):
-        part = _legacy_version_replacement_map.get(part, part)
+        part_y = _legacy_version_replacement_map.get(part, part)
```
Comment (MarcoGorelli) [inline pandas/util/version/__init__.py:256]: why `_y`? what does `y` refer to?

File: scripts/validate_docstrings.py [hunk 3]
```diff
@@ -103,33 +103,33 @@ def get_api_items(api_doc_fd):
     previous_line = current_section = current_subsection = ""
     position = None
     for line in api_doc_fd:
-        line = line.strip()
-        if len(line) == len(previous_line):
-            if set(line) == set("-"):
+        line_strip = line.strip()
```
Comment (MarcoGorelli) [inline scripts/validate_docstrings.py:106]: this one's nice, but perhaps let's call it `line_stripped`?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for working on this

to be honest, I'm no longer sold on enabling this code, having seen the changes it entails it probably makes things less clear more often than not. this is my fault for having suggested the issue, sorry about that

the `line_stripped` change one is nice though, that's probably an improvement

I'd suggest we:
- keep the `line_stripped` change
- revert the rest
and then we can get this in

**Comment by ErdiTk** [comment]
Hi @MarcoGorelli, I just pushed an update of the PR. I renamed all the variables by setting a more intuitive variable name. Please let me know if it's fine for you. In that case I will continue like this also for the other sections.

**Comment by ErdiTk** [comment]
> thanks for working on this
> 
> to be honest, I'm no longer sold on enabling this code, having seen the changes it entails it probably makes things less clear more often than not. this is my fault for having suggested the issue, sorry about that
> 
> the `line_stripped` change one is nice though, that's probably an improvement
> 
> I'd suggest we:
> 
>     * keep the `line_stripped` change
> 
>     * revert the rest
>       and then we can get this in

Oh no problem, no need to say sorry. To be fair I should've thought a bit more on the variable names. Will wait for a feedback on the last push before continuing.

File: web/pandas_web.py [hunk 4]
```diff
@@ -430,8 +430,8 @@ def main(
                 content = extend_base_template(body, context["main"]["base_template"])
             context["base_url"] = "".join(["../"] * os.path.normpath(fname).count("/"))
             content = jinja_env.from_string(content).render(**context)
-            fname = os.path.splitext(fname)[0] + ".html"
-            with open(os.path.join(target_path, fname), "w") as f:
+            fname_splitted = os.path.splitext(fname)[0] + ".html"
+            with open(os.path.join(target_path, fname_splitted), "w") as f:
```
Comment (MarcoGorelli) [inline web/pandas_web.py:434]: I think  this is just about replacing the extension with `html`, so maybe `fname_html` would be better?

File: pandas/_config/config.py [hunk 5]
```diff
@@ -158,8 +158,8 @@ def _set_option(*args, **kwargs) -> None:
             o.validator(v)
 
         # walk the nested dict
-        root, k = _get_root(key)
-        root[k] = v
+        root, k_root = _get_root(key)
+        root[k_root] = v
```
Comment (MarcoGorelli) [inline pandas/_config/config.py:162]: I'm not familiar enough with this part of the code to tell if this new name makes sense, I'll take another look next week

(or, if you've already understood it, feel free to explain)

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for updating, this does look better!

just left a couple of comments

File: pandas/_config/config.py (hunk 5 above)
Comment (ErdiTk) [reply pandas/_config/config.py:162]: I see this method is called from the option context, for example by running:
```
from pandas import option_context
with option_context('display.max_rows', 10, 'display.max_columns', 5):
    pass
```
The `args*` in the call of `option_context` will be passed to _set_option two at the time. From what I see the `_get_root(key)` seems to split the first element passed (in my example above could be `"display.max_rows"`) by dot. Then returns them in a root and key format. From what I understood the `k` element return is associated to the root, that's why I called it `k_root`.

Let me know if I wasn't clear or I made a mistake in understanding the function.

File: web/pandas_web.py (hunk 4 above)
Comment (ErdiTk) [reply web/pandas_web.py]: Agree, just updated name.

File: pandas/_config/config.py (hunk 5 above)
Comment (MarcoGorelli) [reply pandas/_config/config.py:162]: ok thanks for explaining!

**Comment by MarcoGorelli** [comment]
looks like most tests are failing, could you check what it might be please?

**Comment by ErdiTk** [comment]
> looks like most tests are failing, could you check what it might be please?

The line "pandas/core/*" was repeated twice in the pyproject.toml file. I didn't see it in the merge conflict. Now should be fine.

**Comment by ErdiTk** [comment]
The error seems to be connected to my changes in the `scripts/no_bool_in_generic.py` file. I will investigate and fix it today.

**Comment by ErdiTk** [comment]
Hi @MarcoGorelli, I think I completed all the requested changes. Please let me know if I missed something. Just updated the branch with main and fixed the bug I introduced.

**Comment by ErdiTk** [comment]
@MarcoGorelli Sorry I have been away in the last few days. I just pushed the last merge with the main branch. Ideally there should be no more errors. If that is the case do you need any further update of the branch?

File: scripts/validate_docstrings.py [hunk 6]
```diff
@@ -103,33 +103,33 @@ def get_api_items(api_doc_fd):
...
                 position = "items"
                 continue
 
         if position == "items":
-            if line == "":
+            if line_stripped == "":
                 position = None
                 continue
-            item = line.strip()
+            item = line_stripped.strip()
```
Comment (MarcoGorelli) [inline scripts/validate_docstrings.py:132]: lol, the improved name you've chosen actually makes it clear that this line is probably unnecessary 😄 if you wanted to clean this up in a follow-up, that'd be welcome

**Review by MarcoGorelli — APPROVED** [review]
Looks good to me, hopefully this rule will help save some future bugs

Merging then (apologies for the delay, was away last week)

**Pull request merged**